text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1251 @@
+2:03:19 PM | DELETE_IN_PROGRESS   | AWS::DynamoDB::Table                | MyStack/MyTable794EDED1/Resource
+2:03:21 PM | DELETE_COMPLETE      | Custom::DynamoDBReplica             | MyStack/MyTableReplicauswest285A33668/Resource
+2:03:22 PM | DELETE_COMPLETE      | AWS::IAM::ManagedPolicy             | MyStack/MyTableManagedPolicyXXXXXXX/Resource
+2:03:22 PM | DELETE_COMPLETE      | AWS::CloudFormation::NestedStack    | MyStack/ReplicaProviderNestedStack/Resource
+2:03:19 PM | DELETE_COMPLETE      | AWS::DynamoDB::Table                | MyStack/MyTable794EDED1/Resource
+
+✅  Stack deployment complete
+```
+
+> For CI/CD use cases, you can streamline deployments by using the `--require-approval never`
+> option with the CDK CLI to bypass manual approval prompts.
+
+### Drift Detection Failure Example
+
+If a CloudFormation drift is detected, the CLI will show:
+
+```s
+Performing construct upgrade validation...
+
+⚠️  Note: This is an experimental feature
+
+Validations:
+❌ CloudFormation drift detected
+  │ Resource: AWS::DynamoDB::Table MyTable794EDED1
+  │ Actual changes found:
+  └   • BillingMode: "PAY_PER_REQUEST" (expected: "PROVISIONED")
+✅ Change set analysis complete
+
+Error: Construct upgrade validation failed
+Please resolve the drift before proceeding with the upgrade.
+For more information about handling drift, visit: 'https://docs.aws.amazon.com/cdk/v2/guide/drift'
+```
+
+### Change Set Validation Failure Example
+
+**Example 1:** If the change set analysis detects the new table is not being imported (by setting the `TableName` property
+on the construct), the CLI will show:
+
+```s
+Performing construct upgrade validation...
+
+⚠️  Note: This is an experimental feature
+
+Validations:
+✅ No CloudFormation drift detected
+❌ Change set analysis failed
+  │ Resources: MyGlobalTable794EDED1 (AWS::DynamoDB::GlobalTable)
+  │ Change Set:
+  └  • Action: 'Add' (expected: 'Import')
+
+Error: Construct upgrade validation failed
+```
+
+**Example 2:** If the change set analysis detects the existing table does not have policy action `Retain` (by setting
+the `RemovalPolicy` on existing table construct), the CLI will show:
+
+```s
+Performing construct upgrade validation...
+
+⚠️  Note: This is an experimental feature
+
+Validations:
+✅ No CloudFormation drift detected
+❌ Change set analysis failed
+  │ Resource: MyTable794EDED1 (AWS::DynamoDB::Table)
+  └   • RemovalPolicy: 'Delete' (expected: 'Retain')
+
+Error: Construct upgrade validation failed
+```
+
+**Example 3:** If the change set analysis detects the Replica custom resource does not have policy action `Retain`
+(by setting the `ReplicaRemovalPolicy` on existing table construct), the CLI will show:
+
+```s
+Performing construct upgrade validation...
+
+⚠️  Note: This is an experimental feature
+
+Validations:
+✅ No CloudFormation drift detected
+❌ Change set analysis failed
+  │ Resources: MyTableReplicauswest285A33668 (Custom::DynamoDBReplica)
```

Comment:
Did you validate that retaining the custom resource while removing the backing lambda function is possible? CloudFormation doesn't reject such a change?